server/src/test/java/org/elasticsearch/index/translog/TranslogTests.java

```diff
@@ -1507,6 +1512,193 @@ public void testSnapshotFromStreamInput() throws IOException {
         assertEquals(ops, readOperations);
     }
 
+    public void testSnapshotCurrentHasUnexpectedOperationsForTrimmedOperations() throws Exception {
+        long pt = primaryTerm.get();
+        int extraDocs = randomIntBetween(10, 15);
+        for (int op = 0; op < extraDocs; op++) {
+            String ascii = randomAlphaOfLengthBetween(1, 50);
+            Translog.Index operation = new Translog.Index("test", "" + op, op, pt, ascii.getBytes("UTF-8"));
+            translog.add(operation);
+        }
+        try {
+            translog.trimOperations(pt - 1, 0);
+            fail();
+        } catch (AssertionError e) {
+            assertThat(e.getMessage(),
+                is("current should not have any operations with seq#:primaryTerm [0:" + pt + "] >= 0:" + (pt - 1)));
+        }
+
+        translog.rollGeneration();
+
+        // it is possible to trim after generation rollover
+        translog.trimOperations(pt - 1, 0);
+    }
+
+    public void testSnapshotTrimmedOperations() throws Exception {
+        int translogOperations = 0;
+        int skipped = 0;
+        boolean wasRollover = false;
+        Set<Long> expectedSeqNo = new HashSet<>();
+
+        for(int attempt = 0, maxAttempts = randomIntBetween(3, 10); attempt < maxAttempts; attempt++) {
```

Comment:
I think this test got too complicate then what it needs to be. I propose the following:

Have an initial loop that:
 - adds operations to the translog. Sometime new ones with current term, sometimes new ones with older term.
 - within the loop randomly rollover generations and increase the current term.
 - adds every operation it adds to the translog to a list of operations.

- Take the list operations produced by the loop and sort it by seq# and also remove operations that have been super seeded by another operation with the same seq# but a higher term.

- Test multilpe trimming in a loop:
     - select a random trimming number and trim the translog.
     - manually perform the same trimming on the list op operations that we have.
     - compare the translog snapshot output to to the exepected list.

I think this will be simpler to follow and understand.